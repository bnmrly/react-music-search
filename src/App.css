@import url('https://fonts.googleapis.com/css?family=Montserrat:200,400,600&display=swap');

body {
  font-family: 'Montserrat', sans-serif;
  background-color: #eeecec;
}

#root {
  padding: 1.5rem 2rem;
}

.top-albums__app-heading {
  margin-top: 0;
  margin-bottom: 1.5rem;
}

.top-albums__form {
  margin-bottom: 1.5rem;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

@media screen and (min-width: 599px) {
  .top-albums__form {
    flex-direction: row;
    align-items: center;
  }
}

.top-albums__input {
  text-transform: capitalize;
  background-color: #eeecec;
  padding: 0.25rem;
  margin-bottom: 1.25rem;
}

@media screen and (min-width: 599px) {
  .top-albums__input {
    margin-bottom: 0;
  }
}

.top-albums__btn {
  background-color: transparent;
  height: 2rem;
  padding-right: 2rem;
  padding-left: 2rem;
  cursor: pointer;
  border: 1px solid rgb(153, 234, 212);
  border-radius: 50px;
}

@media screen and (min-width: 599px) {
  .top-albums__btn {
    margin-left: 1rem;
  }
}

.top-albums__header {
  margin-bottom: 1.5rem;
}

.top-albums__search-results {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  border: 1px solid rgb(234, 153, 214);
  border-radius: 0.5rem;
  padding: 1.5rem;
}

.top-albums__card {
  display: flex;
  flex-direction: column;
  width: 200px;
  height: auto;
  padding: 1rem;
  border: 1px solid rgb(153, 234, 212);
  border-radius: 0.5rem;
  margin: 0.5rem;
}

.top-albums__title {
  margin: 0;
  padding-top: 0.75rem;
}
