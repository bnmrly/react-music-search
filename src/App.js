import React, { Fragment, useState, useEffect } from 'react';
import axios from 'axios';

function App() {
  const [data, setData] = useState({ popularAlbums: [] });
  const [query, setQuery] = useState('miles davis');
  const [url, setUrl] = useState(
    `http://ws.audioscrobbler.com/2.0/?method=artist.gettopalbums&artist=${query}&api_key=adf0f28675fbabee9c5c0fb4bede5e80&format=json`
  );

  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      setIsError(false);
      setIsLoading(true);

      try {
        const result = await axios(url);
        const popularAlbums = result.data.topalbums.album;
        console.log(popularAlbums);
        setData(popularAlbums);
      } catch (error) {
        setIsError(true);
      }
      setIsLoading(false);
    };
    fetchData();
  }, [url]);

  return (
    <Fragment>
      <h1>Search for your favourite artist's top albums</h1>
      <form
        onSubmit={event => {
          setUrl(
            `http://ws.audioscrobbler.com/2.0/?method=artist.gettopalbums&artist=${query}&api_key=adf0f28675fbabee9c5c0fb4bede5e80&format=json`
          );
          event.preventDefault();
        }}
      >
        <input
          type="text"
          value={query}
          onChange={event => setQuery(event.target.value)}
        />
        <button type="submit">Search</button>
      </form>

      {isError && <div>Something went wrong ...</div>}

      {isLoading ? (
        <div>Loading ...</div>
      ) : (
        // refactor this check
        <Fragment>
          <div>
            {data.length && <h1>Top albums by {data[0].artist.name}</h1>}
            {data.length > 0 &&
              data.map((el, index) => {
                return el.image.map(img => {
                  return img.size === 'large' ? (
                    <div className="topAlbum-info" key={index}>
                      <img src={img['#text']} alt="" />
                      <h4 key={index}>{el.name}</h4>
                    </div>
                  ) : null;
                });
              })}
          </div>
        </Fragment>
      )}
    </Fragment>
  );
}
export default App;
