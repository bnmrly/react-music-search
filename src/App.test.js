import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import App from './App';
import mockAxios from 'axios';

// beforeEach(() => {
// });

afterEach(() => {
  mockAxios.get.mockClear();
});

it('renders the h1 correctly', () => {
  const { getByText } = render(<App />);
  expect(
    getByText("Search for your favourite artist's top albums")
  ).toBeInTheDocument();
});

it('axios get is called once with correct url and Loading... is displayed', () => {
  // arrange

  //  mockAxios.get.mockImplementationOnce can be used to change api call for a specific test instead of using the imported function
  mockAxios.get.mockImplementationOnce(() => {
    Promise.resolve({
      data: { topalbums: { album: [] } }
    });
  });

  // act

  // const { getByText } = render(<App />);

  // look at calling act and get by text - maybe

  let fn;
  act(() => {
    fn = render(<App />).getByText;
  });

  // assert
  expect(mockAxios.get).toHaveBeenCalledTimes(1);
  expect(mockAxios.get).toHaveBeenCalledWith(
    'https://ws.audioscrobbler.com/2.0/?method=artist.gettopalbums&artist=sonic youth&api_key=adf0f28675fbabee9c5c0fb4bede5e80&format=json'
  );

  expect(fn('Loading ...')).toBeInTheDocument();
});
