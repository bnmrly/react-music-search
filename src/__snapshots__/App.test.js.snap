// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <h1
    className="top-albums__app-heading"
  >
    Search for your favourite artist's top albums
  </h1>,
  <div
    className="top-albums__search-container"
  >
    <form
      className="top-albums__form"
      onSubmit={[Function]}
    >
      <input
        className="top-albums__input"
        onChange={[Function]}
        type="text"
        value="sonic youth"
      />
      <button
        className="top-albums__btn"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>,
  <div
    className="top-albums__search-results"
  />,
]
`;
